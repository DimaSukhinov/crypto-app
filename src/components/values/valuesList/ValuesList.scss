.values-list {

  &__rank, &__changes {
    min-width: 65px;
  }

  &__name, &__price, &__symbol {
    min-width: 130px;
  }

  @media (max-width: 630px) {
    &__name, &__changes {
      display: none;
    }
  }

  &__value {
    display: flex;
    justify-content: space-between;
    align-items: center;
    padding: 15px;
    cursor: pointer;

    &:hover {
      background-color: #edeef0;
    }
  }
}
