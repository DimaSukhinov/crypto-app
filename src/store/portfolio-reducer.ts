const initialState: PortfolioType[] = []

export const portfolioReducer = (state: PortfolioType[] = initialState, action: ActionsType): PortfolioType[] => {
    switch (action.type) {
        case 'SET-PORTFOLIO':
            return action.portfolio
        case 'ADD-TO-PORTFOLIO':
            const values = [...state, {
                id: action.id,
                name: action.name,
                price: action.price,
                valueCount: action.valueCount
            }]
            localStorage.setItem('portfolio', JSON.stringify(values));
            return values
        case 'REMOVE-FROM-PORTFOLIO':
            return state.filter(p => p.id !== action.id)
        default:
            return state
    }
}

export const setPortfolioAC = (portfolio: any) => {
    return {type: 'SET-PORTFOLIO', portfolio} as const
}
export const addToPortfolioAC = (id: string, name: string, price: number, valueCount: number) => {
    return {type: 'ADD-TO-PORTFOLIO', id, name, price, valueCount} as const
}
export const removeFromPortfolioAC = (id: string) => {
    return {type: 'REMOVE-FROM-PORTFOLIO', id} as const
}

type ActionsType =
    ReturnType<typeof setPortfolioAC>
    | ReturnType<typeof addToPortfolioAC>
    | ReturnType<typeof removeFromPortfolioAC>

export type PortfolioType = {
    id: string
    name: string
    price: number
    valueCount: number
}
