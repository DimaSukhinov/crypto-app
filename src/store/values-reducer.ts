const initialState: ValuesType[] = []

export const valuesReducer = (state: ValuesType[] = initialState, action: ActionsType): ValuesType[] => {
    switch (action.type) {
        case 'SET-VALUES':
            return action.values.map(v => {
                return {...v}
            })
        default:
            return state
    }
}

export const setValuesAC = (values: ValuesType[]) => {
    return {type: 'SET-VALUES', values} as const
}

type ActionsType = ReturnType<typeof setValuesAC>

export type ValuesType = {
    id: string
    rank: string
    symbol: string
    name: string
    supply: string
    maxSupply: string | null
    marketCapUsd: string
    volumeUsd24Hr: string
    priceUsd: string
    changePercent24Hr: string
    vwap24Hr: string
    explorer: string
}
