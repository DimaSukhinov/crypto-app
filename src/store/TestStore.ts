import { combineReducers, createStore } from 'redux';
import { ValueType } from './values-reducer';
import { PortfolioType } from './portfolio-reducer';

export const values = [
  {
    id: 'bitcoin', rank: '1', changePercent24Hr: '3.1367184772434482', explorer: 'https://blockchain.info/',
    marketCapUsd: '452445368357.8622563021714334', maxSupply: '21000000.0000000000000000', name: 'Bitcoin',
    priceUsd: '23679.9797033025578357', supply: '19106662.0000000000000000', symbol: 'BTC',
    volumeUsd24Hr: '17521511686.5773940471774064', vwap24Hr: '23710.9068155355389843',
  },
  {
    id: 'ethereum', rank: '2', changePercent24Hr: '3.1367184772434482', explorer: 'https://blockchain.info/',
    marketCapUsd: '452445368357.8622563021714334', maxSupply: '21000000.0000000000000000', name: 'Ethereum',
    priceUsd: '23679.9797033025578357', supply: '19106662.0000000000000000', symbol: 'BTC',
    volumeUsd24Hr: '17521511686.5773940471774064', vwap24Hr: '23710.9068155355389843',
  },
];

export const portfolio: PortfolioType[] = [
  { id: 'bitcoin', name: 'Bitcoin', price: 2000, valueCount: 20 },
  { id: 'ethereum', name: 'Ethereum', price: 2000, valueCount: 20 },
];

const ValuesState: ValueType[] = [
  {
    'id': 'bitcoin',
    'rank': '1',
    'symbol': 'BTC',
    'name': 'Bitcoin',
    'supply': '19109256.0000000000000000',
    'maxSupply': '21000000.0000000000000000',
    'marketCapUsd': '439598186222.6843028478185096',
    'volumeUsd24Hr': '9622382958.4556673270153307',
    'priceUsd': '23004.4637123854692641',
    'changePercent24Hr': '-3.4084952567069450',
    'vwap24Hr': '23579.4890013699165648',
    'explorer': 'https://blockchain.info/',
  },
  {
    'id': 'ethereum',
    'rank': '2',
    'symbol': 'ETH',
    'name': 'Ethereum',
    'supply': '121784637.9365000000000000',
    'maxSupply': null,
    'marketCapUsd': '200976786498.8091803887365518',
    'volumeUsd24Hr': '6534517118.1570715975822179',
    'priceUsd': '1650.2638584318075930',
    'changePercent24Hr': '-3.8175749864566575',
    'vwap24Hr': '1696.1297775541851926',
    'explorer': 'https://etherscan.io/',
  },
  {
    'id': 'tether',
    'rank': '3',
    'symbol': 'USDT',
    'name': 'Tether',
    'supply': '66197977707.2567140000000000',
    'maxSupply': null,
    'marketCapUsd': '66250453592.6509238582810531',
    'volumeUsd24Hr': '20837775336.8158968284416711',
    'priceUsd': '1.0007927113064733',
    'changePercent24Hr': '0.1010078632835212',
    'vwap24Hr': '1.0003521044485800',
    'explorer': 'https://www.omniexplorer.info/asset/31',
  },
  {
    'id': 'usd-coin',
    'rank': '4',
    'symbol': 'USDC',
    'name': 'USD Coin',
    'supply': '54399430698.1528200000000000',
    'maxSupply': null,
    'marketCapUsd': '54492359288.5810234021565036',
    'volumeUsd24Hr': '1442267903.9112855456361948',
    'priceUsd': '1.0017082640247439',
    'changePercent24Hr': '0.1224038365462283',
    'vwap24Hr': '0.9993086427840540',
    'explorer': 'https://etherscan.io/token/0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48',
  },
  {
    'id': 'binance-coin',
    'rank': '5',
    'symbol': 'BNB',
    'name': 'BNB',
    'supply': '166801148.0000000000000000',
    'maxSupply': '166801148.0000000000000000',
    'marketCapUsd': '47284799962.0041650982695676',
    'volumeUsd24Hr': '370241395.7990609671005731',
    'priceUsd': '283.4800631108615937',
    'changePercent24Hr': '-3.4432572668922072',
    'vwap24Hr': '288.5517843856753446',
    'explorer': 'https://etherscan.io/token/0xB8c77482e45F1F44dE1745F52C74426C631bDD52',
  },
];

const PortfolioState: PortfolioType[] = [
  { id: 'bitcoin', name: 'Bitcoin', price: 2000, valueCount: 20 },
  { id: 'ethereum', name: 'Ethereum', price: 2000, valueCount: 20 },
];

const rootReducer = combineReducers({
  values: (state: ValueType[] = ValuesState): ValueType[] => {
    return state;
  },
  portfolio: (state: PortfolioType[] = PortfolioState): PortfolioType[] => {
    return state;
  },
});

export const testStore = createStore(rootReducer);
