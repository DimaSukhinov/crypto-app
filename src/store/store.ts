import {combineReducers, createStore} from 'redux';
import {TypedUseSelectorHook, useSelector} from 'react-redux';
import {valuesReducer} from './values-reducer';
const rootReducer = combineReducers({
    values: valuesReducer,
})

export const store = createStore(rootReducer)

export type AppRootStateType = ReturnType<typeof rootReducer>

export const useAppSelector: TypedUseSelectorHook<AppRootStateType> = useSelector

// @ts-ignore
window.store = store
