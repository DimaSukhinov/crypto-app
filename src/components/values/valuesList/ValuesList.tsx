import React, { useCallback } from 'react';
import './ValuesList.scss';
import { ValueType } from '../../../store/values-reducer';
import { Button } from '../../common/button/Button';

type ValueListPropsType = {
  currentPageValues: ValueType[]
  navigateToValue: (id: string) => void
  openAddModal: (id: string) => void
}

export const ValuesList = React.memo(({ navigateToValue, currentPageValues, openAddModal }: ValueListPropsType) => {

  const openValuePage = useCallback((id: string) => () => navigateToValue(id), [navigateToValue]);

  const onOpenAddModal = useCallback((id: string) => () => openAddModal(id), [openAddModal]);

  return (
    <div className='values-list' data-testid={'values-list'}>
      {
        currentPageValues.map(v => <div key={v.name} className={'values-list__value'} onClick={openValuePage(v.id)}
                                        data-testid={'value-list-elem'}>
          <div className={'values-list__rank'}>
            {v.rank}
          </div>
          <div className={'values-list__symbol'}>
            {v.symbol}
          </div>
          <div className={'values-list__name'}>
            {v.name}
          </div>
          <div className={'values-list__price'}>
            {+v.priceUsd > 1 ? +(+v.priceUsd).toFixed(2) : +(+v.priceUsd).toFixed(5)} $
          </div>
          <div className={'values-list__changes'} style={{ color: +v.changePercent24Hr > 0 ? 'green' : 'red' }}>
            {+(+v.changePercent24Hr).toFixed(2)}%
          </div>
          <Button onClickHandler={onOpenAddModal(v.id)}>Add</Button>
        </div>)
      }
    </div>
  );
});
