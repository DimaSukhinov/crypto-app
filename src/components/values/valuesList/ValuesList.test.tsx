import React from 'react';
import {render, screen} from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import {ValuesList} from './ValuesList';
import {values} from '../../../store/TestStore';

test('renders values list components', () => {
    render(
        <ValuesList navigateToValue={jest.fn()} currentPageValues={values} openAddModal={jest.fn()}/>
    )

    expect(screen.getByTestId('values-list')).toBeInTheDocument()
    expect(screen.getByTestId('values-list')).toMatchSnapshot()

    const valueElem = screen.getAllByTestId('value-list-elem')

    expect(valueElem).toHaveLength(2) // expect render of 2 values
    expect(valueElem[0]).toHaveTextContent(values[0].name)
});
