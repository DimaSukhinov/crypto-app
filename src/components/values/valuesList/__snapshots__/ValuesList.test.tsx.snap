// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Values list renders values list components 1`] = `
<div
  class="values-list"
  data-testid="values-list"
>
  <div
    class="values-list__value"
    data-testid="value-list-elem"
  >
    <div
      class="values-list__rank"
    >
      1
    </div>
    <div
      class="values-list__symbol"
    >
      BTC
    </div>
    <div
      class="values-list__name"
    >
      Bitcoin
    </div>
    <div
      class="values-list__price"
    >
      23679.98
       $
    </div>
    <div
      class="values-list__changes"
      style="color: green;"
    >
      3.14
      %
    </div>
    <div
      class="button"
    >
      Add
    </div>
  </div>
  <div
    class="values-list__value"
    data-testid="value-list-elem"
  >
    <div
      class="values-list__rank"
    >
      2
    </div>
    <div
      class="values-list__symbol"
    >
      BTC
    </div>
    <div
      class="values-list__name"
    >
      Ethereum
    </div>
    <div
      class="values-list__price"
    >
      23679.98
       $
    </div>
    <div
      class="values-list__changes"
      style="color: green;"
    >
      3.14
      %
    </div>
    <div
      class="button"
    >
      Add
    </div>
  </div>
</div>
`;

exports[`renders values list components 1`] = `
<div
  class="values-list"
  data-testid="values-list"
>
  <div
    class="values-list__value"
    data-testid="value-list-elem"
  >
    <div
      class="values-list__rank"
    >
      1
    </div>
    <div
      class="values-list__symbol"
    >
      BTC
    </div>
    <div
      class="values-list__name"
    >
      Bitcoin
    </div>
    <div
      class="values-list__price"
    >
      23679.98
       $
    </div>
    <div
      class="values-list__changes"
      style="color: green;"
    >
      3.14
      %
    </div>
    <div
      class="button"
    >
      Add
    </div>
  </div>
  <div
    class="values-list__value"
    data-testid="value-list-elem"
  >
    <div
      class="values-list__rank"
    >
      2
    </div>
    <div
      class="values-list__symbol"
    >
      BTC
    </div>
    <div
      class="values-list__name"
    >
      Ethereum
    </div>
    <div
      class="values-list__price"
    >
      23679.98
       $
    </div>
    <div
      class="values-list__changes"
      style="color: green;"
    >
      3.14
      %
    </div>
    <div
      class="button"
    >
      Add
    </div>
  </div>
</div>
`;
